Clock Skew(?) can cause the distruptor thread to crash

Please answer these questions before submitting a bug report.

### What version of OpenCensus are you using?
0.24.0, but looks to be present in all versions

### What JVM are you using (`java -version`)?
1.8.0

Occasionally, at process start, we'll see this error in our logs:

```
Exception in thread "OpenCensus.Disruptor-0" java.lang.RuntimeException: java.lang.IllegalArgumentException: Current time must be within or after the last bucket.
	at com.lmax.disruptor.FatalExceptionHandler.handleEventException(FatalExceptionHandler.java:45)
	at com.lmax.disruptor.dsl.ExceptionHandlerWrapper.handleEventException(ExceptionHandlerWrapper.java:18)
	at com.lmax.disruptor.BatchEventProcessor.processEvents(BatchEventProcessor.java:187)
	at com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:125)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: Current time must be within or after the last bucket.
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:141)
	at io.opencensus.implcore.stats.MutableViewData$IntervalMutableViewData.refreshBucketList(MutableViewData.java:308)
	at io.opencensus.implcore.stats.MutableViewData$IntervalMutableViewData.record(MutableViewData.java:262)
	at io.opencensus.implcore.stats.MeasureToViewMap.record(MeasureToViewMap.java:160)
	at io.opencensus.implcore.stats.StatsManager$StatsEvent.process(StatsManager.java:101)
	at io.opencensus.impl.internal.DisruptorEventQueue$DisruptorEventHandler.onEvent(DisruptorEventQueue.java:229)
	at io.opencensus.impl.internal.DisruptorEventQueue$DisruptorEventHandler.onEvent(DisruptorEventQueue.java:222)
	at com.lmax.disruptor.BatchEventProcessor.processEvents(BatchEventProcessor.java:168)
	... 2 more
```

If this happens, all publisher threads eventually will hang forever, since there is no longer a consumer of the LMAX Disruptor queue.  It seems like the cause of this is clock skew causing measurements to occur before the bucket start, and there is even a TODO in the code to handle this:
https://github.com/census-instrumentation/opencensus-java/blob/master/impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java#L307

I'm not sure what the correct behavior should be when the measurement timestamp is before the bucket start though, drop the event possibly?

--- a/impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java
+++ b/impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java
@@ -132,6 +132,7 @@ abstract class MutableViewData {
     @javax.annotation.Nullable
     @Override
     Metric toMetric(Timestamp now, State state) {
+      handleTimeRewinds(now);
       if (state == State.DISABLED) {
         return null;
       }
@@ -166,6 +167,7 @@ abstract class MutableViewData {
 
     @Override
     ViewData toViewData(Timestamp now, State state) {
+      handleTimeRewinds(now);
       if (state == State.ENABLED) {
         return ViewData.create(
             super.view,
@@ -177,6 +179,18 @@ abstract class MutableViewData {
             super.view,
             Collections.<List</*@Nullable*/ TagValue>, AggregationData>emptyMap(),
             ViewData.AggregationWindowData.CumulativeData.create(ZERO_TIMESTAMP, ZERO_TIMESTAMP));
+      }
+    }
+
+    /**
+     * This method attemps to migrate this view into a reasonable state in the event of time going
+     * backwards.
+     */
+    private void handleTimeRewinds(Timestamp now) {
+      if (now.compareTo(start) < 0) {
+        // Time went backwards, physics is broken, forget what we know.
+        tagValueAggregationMap.clear();
+        start = now;
       }
     }
 
@@ -311,6 +325,7 @@ abstract class MutableViewData {
         // - Drop events in the future, keep others within a duration.
         // - Drop all events on skew
         // - Guess at time-skew and "fix" events
+        // - Reset our "start" time to now if necessary.
         buckets.clear();
         shiftBucketList(N + 1, now);
         return;

--- a/impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java
+++ b/impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java
@@ -189,7 +189,7 @@ abstract class MutableViewData {
     private void handleTimeRewinds(Timestamp now) {
       if (now.compareTo(start) < 0) {
         // Time went backwards, physics is broken, forget what we know.
-        tagValueAggregationMap.clear();
+        clearStats();
         start = now;
       }
     }

--- a/impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java
+++ b/impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java
@@ -304,10 +304,17 @@ abstract class MutableViewData {
       }
       Timestamp startOfLastBucket =
           CheckerFrameworkUtils.castNonNull(buckets.peekLast()).getStart();
-      // TODO(songya): decide what to do when time goes backwards
-      checkArgument(
-          now.compareTo(startOfLastBucket) >= 0,
-          "Current time must be within or after the last bucket.");
+      // Time went backwards!  Physics has failed us!  drop everything we know and relearn.
+      // Prioritize:  Report data we're confident is correct.
+      if (now.compareTo(startOfLastBucket) < 0) {
+        // TODO: configurable time-skew handling with options:
+        // - Drop events in the future, keep others within a duration.
+        // - Drop all events on skew
+        // - Guess at time-skew and "fix" events
+        buckets.clear();
+        shiftBucketList(N + 1, now);
+        return;
+      }
       long elapsedTimeMillis = now.subtractTimestamp(startOfLastBucket).toMillis();
       long numOfPadBuckets = elapsedTimeMillis / bucketDuration.toMillis();
 

--- a/impl_core/src/test/java/io/opencensus/implcore/stats/MutableViewDataTest.java
+++ b/impl_core/src/test/java/io/opencensus/implcore/stats/MutableViewDataTest.java
@@ -19,6 +19,19 @@ package io.opencensus.implcore.stats;
 import static com.google.common.truth.Truth.assertThat;
 
 import io.opencensus.common.Timestamp;
+import io.opencensus.implcore.internal.CurrentState;
+import io.opencensus.implcore.tags.TagMapImpl;
+import io.opencensus.metrics.data.AttachmentValue;
+import io.opencensus.stats.Aggregation;
+import io.opencensus.stats.Aggregation.Count;
+import io.opencensus.stats.Aggregation.Distribution;
+import io.opencensus.stats.BucketBoundaries;
+import io.opencensus.stats.Measure.MeasureDouble;
+import io.opencensus.stats.View;
+import io.opencensus.stats.ViewData;
+import io.opencensus.tags.TagKey;
+import java.util.Arrays;
+import java.util.Collections;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
@@ -30,5 +43,54 @@ public class MutableViewDataTest {
   @Test
   public void testConstants() {
     assertThat(MutableViewData.ZERO_TIMESTAMP).isEqualTo(Timestamp.create(0, 0));
+  }
+
+  @Test
+  public void testTimeRewindsOnCountViewNoThrow() {
+    // First we set up some buckets THEN we rewind time for giggles.
+    View tester =
+        View.create(
+            View.Name.create("view"),
+            "Description",
+            MeasureDouble.create("name", "desc", "us"),
+            Count.create(),
+            Collections.singletonList(TagKey.create("KEY")));
+    Timestamp start = Timestamp.create(10000000, 0);
+    Timestamp validPointTime = Timestamp.create(10000010, 0);
+    CurrentState.State state = CurrentState.State.ENABLED;
+    MutableViewData viewData = MutableViewData.create(tester, start);
+    // Create a data points to get thrown away.
+    viewData.record(
+        TagMapImpl.EMPTY, 1.0, validPointTime, Collections.<String, AttachmentValue>emptyMap());
+    // Rewind time and look for explosions.
+    Timestamp thePast = Timestamp.create(0, 0);
+    ViewData result = viewData.toViewData(thePast, state);
+    assertThat(result.getAggregationMap()).isEmpty();
+  }
+
+  @Test
+  public void testTimeRewindsOnDistributionViewNoThrow() {
+    // First we set up some buckets THEN we rewind time for giggles.
+    Aggregation latencyDistribution =
+        Distribution.create(
+            BucketBoundaries.create(Arrays.asList(0.0, 25.0, 100.0, 200.0, 400.0, 800.0, 10000.0)));
+    View tester =
+        View.create(
+            View.Name.create("view"),
+            "Description",
+            MeasureDouble.create("name", "desc", "us"),
+            latencyDistribution,
+            Collections.singletonList(TagKey.create("KEY")));
+    Timestamp start = Timestamp.create(10000000, 0);
+    Timestamp validPointTime = Timestamp.create(10000010, 0);
+    CurrentState.State state = CurrentState.State.ENABLED;
+    MutableViewData viewData = MutableViewData.create(tester, start);
+    // Create a data points to get thrown away.
+    viewData.record(
+        TagMapImpl.EMPTY, 1.0, validPointTime, Collections.<String, AttachmentValue>emptyMap());
+    // Rewind time and look for explosions.
+    Timestamp thePast = Timestamp.create(0, 0);
+    ViewData result = viewData.toViewData(thePast, state);
+    assertThat(result.getAggregationMap()).isEmpty();
   }
 }

